### ISSUE TITLE
JavaScript heap out of memory

### ISSUE THREAD

**Issue by j2l** [axa-group/nlp.js#105]
**Describe the bug**
Out of memory error when training on a [public set](https://www.kaggle.com/c/spooky-author-identification/data). Full error message below.

**To Reproduce**
Download, unzip the [train.csv](https://www.kaggle.com/c/spooky-author-identification/data) file, adapt it to XLS (NER is empty):
NLP:
```
intent | language | utterance
EAP | en | id26305, This process, however, afforded me no means of   ascertaining the dimensions of my dungeon; as I might make its circuit, and   return to the point whence I set out, without being aware of the fact; so   perfectly uniform seemed the wall.
... (19568 lines!)
```

NLG:
```
intent | condition | language | response
EAP |   | en | EAP
MWS |   | en | MWS
HPL |   | en | HPL
```

**Expected behavior**
A trained model saved to model.nlp

**More globally**
Could we have an example of some tests script using XLS?

** Full error message**
```
Training, please wait..

<--- Last few GCs --->

[5192:000FA808]   298302 ms: Mark-sweep 666.6 (736.9) -> 666.6 (736.9) MB, 596.3 / 0.0 ms  allocation failure GC in old space requested
[5192:000FA808]   298985 ms: Mark-sweep 666.6 (736.9) -> 666.6 (720.9) MB, 682.9 / 0.0 ms  last resort GC in old space requested[5192:000FA808]   299666 ms: Mark-sweep 666.6 (720.9) -> 666.6 (720.4) MB, 681.5 / 0.0 ms  last resort GC in old space requested

<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 048961D9 <JSObject>
    1: /* anonymous */(aka /* anonymous */) [E:\disctut\node_modules\node-nlp\lib\nlp\nlp-classifier.js:~195] [pc=0C2E84FB](this=33C0417D <undefined>,srcToken=01E45131 <String[7]: id03416>)
    2: arguments adaptor frame: 3->1
    3: forEach(this=18768569 <JSArray[35184]>)
    4: tokensToNeural [E:\disctut\node_modules\node-nlp\lib\nlp\nlp-classifier.js:195] [bytecode=01F6C4AD offset=153](this=061E773...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: node_module_register
 2: v8::internal::Factory::NewFixedArray
 3: v8::internal::HashTable<v8::internal::SeededNumberDictionary,v8::internal::SeededNumberDictionaryShape>::IsKey
```

**Comment by jesus-seijas-sp** [comment]
Yeah, I know that one :) There are some optimizations that can be done when generating the dataset for brain.js, so the memory consumed is not so big, but also you can increase the heap of node... I got out of heap memory in node when playing around with tensorflow also, with big datasets.
You can change your npm start with : node --max-old-space-size=4096 .
Also, this training set is so big (19580 phrases). Usually cloud services are limited, example microsoft limit to 15k utterances. This is becaue the matrix for features/intents will be so big...

So if you want to focus in this exercise, I can help you, but first things that I suggest are:
- Increase the heap memory (command I passed allows 4GB)
- Put a target timeout for intent in the Binary Relevance Neural Network:
https://github.com/axa-group/nlp.js/blob/master/lib/classifiers/binary-neural-network-classifier.js 
here you can see the configuration. I suggest 1 minute per intent:

 this.settings.config = {
        activation: 'leaky-relu',
        hiddenLayers: [],
        learningRate: 0.1,
        errorThresh: 0.0005,
        timeout: 60*1000,
      };

I got it trained with some changes with a dataset of 12k utterance... I never tried 19K utterance before, but I can give it a try.

**Comment by j2l** [comment]
Thank you for giving so many insights!
I'll try both but don't worry, I wanted to compare to other tools and simply copy pasted the set, I don't need 19k for now. May be later though. 

I'm studying nlu-benchmark/index.js
 Could we have an example of some tests script using XLS?
I think it will help people understand how a test is run (me included).
If it's already done or easy to do. Don't want to waste your time 😄

**Comment by j2l** [comment]
`node server.js --max-old-space-size=8192` and        ` timeout: 30*1000,` didn't make it

**Comment by jesus-seijas-sp** [comment]
Hi @hayksaakian , the case provided by j2l is over 15.000 utterances to be classified by intents, so is a process of the AI training that consumes memory. 
But your case is about entity recognition, your provided files are supposed to be entities and not intents, so no AI process is in the middle, so it seems very odd that you get into a heap out of memory...

Are you using them as NER? or in the other hand are you trying to train them in the NLU?

**Comment by jesus-seijas-sp** [comment]
Solved in version 3.7.0.
